File: frontend/src/types/blueprint.ts
export interface CellResponse {
  id: number;
  blueprint_id: number;
  row_idx: number;
  col_idx: number;
  bead_code: string | null;
  pixel_color: string | null;
}

export interface CellUpdateRequest {
  id: number;
  bead_code: string;
}

export interface CellUpdateBatch {
  cells: CellUpdateRequest[];
}

export interface Blueprint {
  id: number;
  name: string | null;
  original_filename: string | null;
  grid_rows: number;
  grid_cols: number;
  status: BlueprintStatus;
  created_at: string;
}

export interface BlueprintDetail extends Blueprint {
  cells: CellResponse[];
}

export interface UploadResponse {
  id: number;
  status: BlueprintStatus;
  message: string;
}

export interface StatusResponse {
  id: number;
  status: BlueprintStatus;
  progress: string | null;
}

export type BlueprintStatus = 'processing' | 'ready' | 'error';

export type { CellResponse as Cell };

File: backend/app/schemas/blueprint.py
from pydantic import BaseModel, ConfigDict, Field
from typing import Optional, List
from datetime import datetime


class CellResponse(BaseModel):
    id: int
    blueprint_id: int
    row_idx: int
    col_idx: int
    bead_code: Optional[str] = None
    pixel_color: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class CellUpdateRequest(BaseModel):
    id: int
    bead_code: str = Field(..., min_length=1, max_length=10)


class CellUpdateBatch(BaseModel):
    cells: List[CellUpdateRequest]


class BlueprintCreate(BaseModel):
    name: Optional[str] = None


class BlueprintResponse(BaseModel):
    id: int
    name: Optional[str] = None
    original_filename: Optional[str] = None
    grid_rows: int
    grid_cols: int
    valid_codes: Optional[str] = None
    status: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BlueprintDetailResponse(BlueprintResponse):
    cells: List[CellResponse] = []


class UploadResponse(BaseModel):
    id: int
    status: str
    message: str = "Blueprint uploaded and queued for parsing"


class StatusResponse(BaseModel):
    id: int
    status: str
    progress: Optional[str] = None
